File: spec/factories/assessments.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :assessment, class: Assessments::Common do
    sequence(:name) { |i| "assessment #{i}" }
    description { Faker::Lorem.characters(number: 5) }
    dimension
    type { Assessments::Common }
    extra { { icon_color: '#845EC2' } }

    trait :with_report do
      after(:create) do |assessment, _evaluator|
        create :report, assessments: [assessment]
      end
    end

    trait :with_same_owner_dimension do
      after(:create) do |assessment, _evaluator|
        if assessment.dimension && assessment.owner.present?
          assessment.dimension.update!(owner: assessment.owner)
        end
      end
    end

    factory :hogan_assessment, class: '::Assessments::Hogan' do
      category { Assessment::CATEGORIES[:hogan] }
      type { Assessments::Hogan }
      external_settings { { assessment_id: 'HDS' } }
    end

    trait :iiht do
      category { Assessment::CATEGORIES[:iiht] }
      type { Assessments::Iiht }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :saville do
      category { Assessment::CATEGORIES[:saville] }
      type { Assessments::Saville }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :pearson do
      category { Assessment::CATEGORIES[:pearson] }
      type { Assessments::Pearson }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :mettl do
      category { Assessment::CATEGORIES[:mettl] }
      type { Assessments::Mettl }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :simulation do
      category { Assessment::CATEGORIES[:simulation] }
      type { Assessments::Simulation }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :skillvue do
      category { Assessment::CATEGORIES[:skillvue] }
      type { Assessments::Skillvue }
      external_settings { { assessment_id: '36d0676c454f67d39e67aee0699a2edd' } }
    end

    trait :yoodli do
      category { Assessment::CATEGORIES[:yoodli] }
      type { Assessments::Yoodli }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :mhs do
      category { Assessment::CATEGORIES[:mhs] }
      type { Assessments::Mhs }
      external_settings { { assessment_id: 'assessmentId' } }
    end

    trait :microsite do
      category { Assessment::CATEGORIES[:microsite] }
      type { Assessments::Microsite }
      external_settings { { assessment_id: 'microsite-assessment-id' } }
    end
  end

  factory :assessment_hogan, class: 'Assessments::Hogan' do
    sequence(:name) { |i| "hogan assessment #{i}" }
    extra { { icon_color: '#845EC2' } }
    external_settings { { assessment_id: 'assessmentId' } }
  end
end

File: app/frontend/modules/admin/modules/client/core/assessments.ts
import * as t from 'io-ts'
import ApiAction from 'interfaces/ApiAction'
import { UploadFile } from 'antd/es/upload/interface'
import { createReducer } from '~/utils/redux'

export const ExternalSettingsTR = t.type({
  assessmentId: t.union([t.string, t.null, t.undefined]),
  normId: t.union([t.string, t.null, t.undefined]),
  scheduleConfig: t.union([t.string, t.null, t.undefined]),
  questionMappings: t.union([t.string, t.null, t.undefined]),
  assessmentName: t.union([t.string, t.null, t.undefined]),
  normName: t.union([t.string, t.null, t.undefined]),
})
export type ExternalSettings = t.TypeOf<typeof ExternalSettingsTR>

export const AssessmentTR = t.type({
  id: t.string,
  name: t.string,
  disabled: t.boolean,
  archived: t.boolean,
  deleted: t.boolean,
  defaultLanguage: t.string,
  iconUrl: t.union([t.string, t.null, t.undefined]),
  iconColor: t.union([t.string, t.null, t.undefined]),
  type: t.string,
  category: t.string,
  timing: t.union([t.string, t.null]),
  status: t.union([t.string, t.null]),
  createdAt: t.union([t.string, t.null]),
  updatedAt: t.union([t.string, t.null]),
  createdBy: t.union([t.string, t.null]),
  modifiedBy: t.union([t.string, t.null]),
  enableVideoCheck: t.union([t.boolean, t.null]),
  enableAudioCheck: t.union([t.boolean, t.null]),
  enableNetworkCheck: t.union([t.boolean, t.null]),
  translationsMigrated: t.union([t.boolean, t.null]),
  externalSettings: ExternalSettingsTR,
  icon: t.union([t.string, t.null]),
  tagList: t.array(t.union([t.string, t.null])),
  poster: t.union([t.string, t.null]),
  dimension: t.union([
    t.type({
      id: t.string,
      name: t.string,
    }),
    t.undefined]),
  owner: t.union([
    t.type({
      id: t.string,
      name: t.string,
    }),
    t.undefined]),
  project: t.union([
    t.type({
      id: t.string,
      name: t.string,
    }),
    t.undefined]),
  meta: t.type({
    permissions: t.type({
      manage: t.boolean,
      exportRawResults: t.boolean,
      exportRawFactorScores: t.boolean,
      exportNormedResults: t.boolean,
      externalScores: t.boolean,
    }),
  }),
  linkedAssessment: t.union([
    t.type({
      id: t.string,
      name: t.string,
    }),
    t.undefined, t.null]),
  questions: t.union([
    t.array(t.type({
      id: t.string,
      name: t.string,
    })),
    t.undefined]),
})


export type Assessment = t.TypeOf<typeof AssessmentTR>
export type LinkedAssessment = {
  id: string
  name: string
}

export const CATEGORIES = [
  'psychometric',
  'organisational',
  'case_study',
  'threesixty',
  'assessor_form',
  'lead_assessor_form',
  'agile',
  'hogan',
  'saville',
  'pearson',
  'iiht',
  'meeting',
  'mhs',
]

export const UPDATABLE_CATEGORIES = [
  'psychometric',
  'organisational',
  'case_study',
  'meeting',
]

export const CREATABLE_CATEGORIES = [
  ...UPDATABLE_CATEGORIES,
  'agile',
  'threesixty',
  'assessor_form',
  'lead_assessor_form',
]


export const TYPES = [
  'common',
  'hogan',
  'iiht',
  'pearson',
  'saville',
  'mettl',
  'mhs',
  'simulation',
  'skillvue',
  'yoodli',
  'microsite',
]

export const Schema = {
  type: 'assessments',
  relationships: {
    owner: {
      type: 'clients',
    },
    project: {
      type: 'clients',
    },
    dimension: {
      type: 'dimensions',
    },
    linkedAssessment: {
      type: 'assessments',
    },
    questions: {
      type: 'questions',
    },
  },
}

export type Files = {
  icon: {
    file?: File | UploadFile
  }
  poster: {
    file?: File | UploadFile
  }
}

export const UPLOAD_FILES = 'resource/assessments/UPLOAD_FILES'

export const uploadFiles = (id: string, data: FormData): ApiAction<void> => ({
  type: UPLOAD_FILES,
  request: {
    method: 'put',
    contentType: 'multipart/form-data;',
    url: `/api/v2/administration/assessments/${id}/uploads`,
    body: data,
  },
})


export const reducer = createReducer({}, null)
